### PR TITLE
Recursive option for folderlist

### PR DESCRIPTION
**Pull request by laoneo**
#### Summary of Changes

The folderlist form field got a new option to show the list the folders recursive.
#### Testing Instructions
- Add the following XML string to the file administrator/components/com_content/models/forms/article.xml in the basic fieldset around line 152.

``` xml
    <field name="directory" type="folderlist" directory="images" class="input-xxlarge"
            required="true" hide_none="true" hide_default="true" recursive="true"
            label="COM_FIELDS_FIELD_IMAGELIST_DIRECTORY_LABEL"
            description="COM_FIELDS_FIELD_IMAGELIST_DIRECTORY_DESC">
            <option value=""></option>
    </field>
```
- Open an article in the back end.
- Navigate to the options tab.
#### Expected result

You should see a new select box which lists all folders recursively which are in the images folder.
#### Actual result

Only the folders below the images folders are shown.

Would be great to get that into 3.6 as it is needed for https://github.com/joomla-projects/custom-fields/issues/97.

### CODE REVIEW

File: libraries/joomla/form/fields/folderlist.php [hunk 1]
```diff
@@ -188,7 +201,7 @@ protected function getOptions()
 		}
 
 		// Get a list of folders in the search path with the given filter.
-		$folders = JFolder::folders($path, $this->filter);
+		$folders = JFolder::folders($path, $this->filter, $this->recursive, $this->recursive);
```
Comment (brianteeman) [inline libraries/joomla/form/fields/folderlist.php]: Is it correct that you have this->recursive twice?

Comment (laoneo) [inline libraries/joomla/form/fields/folderlist.php]: It ensures bc.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on ad1dfa32e79d86ad249666a77d642aaedb5baad9<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10722">issues.joomla.org/joomla-cms/10722</a>.</sub>

**Comment by andrepereiradasilva** [comment]
This works as described.
Just one comment (besides brian's comment).

Before patch we have:
```
banners
headers
sampledata
```

After patch i get (notice the slashes):

```
\banners
\headers
\sampledata
\sampledata\fruitshop
\sampledata\parks
\sampledata\parks\animals
\sampledata\parks\landscape
```

So, wouldn't it be more consistent with other tree like views in select boxes (ex: menu items) if we have?:
```
banners
headers
sampledata
- fruitshop
- parks
- - animals
- - landscape
```
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10722">issues.joomla.org/joomla-cms/10722</a>.</sub>

**Comment by brianteeman** [comment]
Not sure as this time it refers to a real physical tree not a virtual one.

**Comment by andrepereiradasilva** [comment]
hum, but also remeber we will get huge values if we put the whole tree

**Comment by brianteeman** [comment]
excellent point

On 5 June 2016 at 13:16, andrepereiradasilva <notifications@github.com>
wrote:

> hum, but also remeber we will get huge values if we put the whole tree
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/10722#issuecomment-223809982>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABPH8Rs2mNpLlu5efZIygUZv0nEvnuKlks5qIr4YgaJpZM4IuGx4>
> .
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

**Comment by laoneo** [comment]
The tree is describing a real path in the Filesystem, so I would show it like that.

I put in the root and the tree was big, but the page got still loaded decently.

**Comment by brianteeman** [comment]
@laoneo I agree and I thought the same at first. But. When you have a very long path eg images/articles/2015/January/finance/blogposts the it probably gets too wide to be usable. Using a tree as proposed resolves that.

**Comment by Bakual** [comment]
It's true that it may become more readable in some cases (in others probably not).
However the code change becomes much more complex. Currently, we just use the API and return what we get.

I would leave this PR with the full path like it currently is, and if we want we can change it to a tree later.

**Comment by andrepereiradasilva** [comment]
ok for me

**Comment by laoneo** [comment]
Can somebody test it please?

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 32c9048b88383ba6028743a48517176952d9a574<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10722">issues.joomla.org/joomla-cms/10722</a>.</sub>

**Comment by andrepereiradasilva** [comment]
I have tested this item :white_check_mark: successfully on 32c9048b88383ba6028743a48517176952d9a574<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10722">issues.joomla.org/joomla-cms/10722</a>.</sub>

**Comment by brianteeman** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/10722">issues.joomla.org/joomla-cms/10722</a>.</sub>

**Pull request merged**